### ISSUE TITLE
Editor crashes when adding many entities with mesh & material components

### ISSUE THREAD

**Issue by galibzon** [o3de/o3de#10699]
**Describe the bug**
I'm putting together a python script that can be used to populate a level with many entities.
The attached  file contains everything that is needed to replicate the crash.

**Assets required**
[goal_9ks_1kd.zip](https://github.com/o3de/o3de/files/9124567/goal_9ks_1kd.zip)

**Steps to reproduce**
1. Unzip the attached zip file and unzip it under the `o3de\AutomatedTesting\Levels` folder.
2. Open the Editor.
3. Load the level called `goal_9ks_1kd`
4. In the `Console` execute the following command:
 `pyRunFile D:\GIT\o3de\AutomatedTesting\Levels\goal_9ks_1kd\Scripts\populate_static_objects.py -m 100 -o 0 -c 100`
5. The above command 

**Expected behavior**
One hundred entities should be added to the terrain, each entity contains shape, physics, mesh and material components.

**Actual behavior**
At random times the Editor will consistently crash with the following stack (In this case the script crashed when adding entity number 60) :
```
>	[Inline Frame] AtomLyIntegration_CommonFeatures.Editor.dll!AZ::EBus<AZ::Render::EditorMaterialSystemComponentNotifications,AZ::Render::EditorMaterialSystemComponentNotifications>::DisconnectInternal(AZ::EBus<AZ::Render::EditorMaterialSystemComponentNotifications,AZ::Render::EditorMaterialSystemComponentNotifications>::Context &) Line 1157	C++
 	AtomLyIntegration_CommonFeatures.Editor.dll!AZ::Internal::NonIdHandler<AZ::Render::EditorMaterialSystemComponentNotifications,AZ::Render::EditorMaterialSystemComponentNotifications,AZ::Internal::EBusContainer<AZ::Render::EditorMaterialSystemComponentNotifications,AZ::Render::EditorMaterialSystemComponentNotifications,0,1>>::BusDisconnect() Line 1417	C++
 	AtomLyIntegration_CommonFeatures.Editor.dll!AZ::Render::EditorMaterialComponentSlot::~EditorMaterialComponentSlot() Line 165	C++
 	AtomLyIntegration_CommonFeatures.Editor.dll!AZStd::vector<AZ::Render::EditorMaterialComponentSlot,AZStd::allocator>::~vector<AZ::Render::EditorMaterialComponentSlot,AZStd::allocator>() Line 341	C++
 	AtomLyIntegration_CommonFeatures.Editor.dll!AZStd::vector<AZStd::vector<AZ::Render::EditorMaterialComponentSlot,AZStd::allocator>,AZStd::allocator>::~vector<AZStd::vector<AZ::Render::EditorMaterialComponentSlot,AZStd::allocator>,AZStd::allocator>() Line 341	C++
 	[External Code]	
 	[Inline Frame] EditorLib.dll!AZ::Entity::Reset() Line 122	C++
 	EditorLib.dll!AZ::Entity::~Entity() Line 97	C++
 	[External Code]	
 	[Inline Frame] EditorLib.dll!std::default_delete<AZ::Entity>::operator()(AZ::Entity *) Line 3120	C++
...
```
Complete stack: 
[editor_crash_python_entities.txt](https://github.com/o3de/o3de/files/9124551/editor_crash_python_entities.txt)

**Screenshots/Video**
Use the attached level

**Found in Branch**
commit 9a15f038807ba8b987c9e689952d9271ef7fd086 (HEAD -> development, upstream/development)
Author: moraaar <moraaar@amazon.com>
Date:   Fri Jul 15 10:12:04 2022 +0100

**Desktop/Device (please complete the following information):**
 - Device: PC 
 - OS: Windows
 - Version 10
 - CPU AMD Ryzen Threadripper 3970X 32-Core Processor    3.70 GHz
 - GPU NVidia RTX 3070
 - Memory 64GB

**Additional context**
I'm using the python script to create a level with 9000+ entities to measure performance.

**Comment by gadams3** [comment]
After doing some investigation, I think it's just coincidence that this is landing in material component code. I think the problems relate to the implementation of PyIdleWaitFrames (idle_wait_frames), which is called from your script and other test util functions. Your script is also saving repeatedly which invokes aspects of the prefab system that may not need to run during your script. If I disable the PyIdleWaitFrames (idle_wait_frames) by not allowing it to execute its own event loop then your script runs perfectly fine. Well, I also deleted this portion...

```
            general.idle_wait_frames(1)
            elapsed_time = time.time() - start_time
            print(f"Added Entity {idx} of {numEntities}. Elapsed {elapsed_time} seconds")
            batch_count = batch_count + 1
            if (saveCount > 0) and (batch_count >= saveCount):
                general.save_level()
                print("Level saved")
                batch_count = 0
```

**Comment by galibzon** [comment]
Interesting. I added the call to save_level() because it was crashing every 30+ entities so I can save and restart from where it last landed before crashing. So you are saying you disabled the logic in the C++ code of PyIdleWaitFrames () and you noticed the crashes stop occurring?

**Comment by jackalbe** [comment]
If this is part of all the call stacks, maybe we should ask the Game Object teams what might be happening? 

**EditorLib.dll!AzToolsFramework::Prefab::JsonInstanceSerializer::Reload()**

`	
EditorLib.dll!AzToolsFramework::Prefab::JsonInstanceSerializer::Reload(const rapidjson_ly::GenericValue<rapidjson_ly::UTF8<char>,rapidjson_ly::MemoryPoolAllocator<rapidjson_ly::CrtAllocator>> & inputValue, AZ::JsonDeserializerContext & context, AzToolsFramework::Prefab::Instance * instance, rapidjson_ly::GenericDocument<rapidjson_ly::UTF8<char>,rapidjson_ly::MemoryPoolAllocator<rapidjson_ly::CrtAllocator>,rapidjson_ly::CrtAllocator> patches, AZ::JsonSerializationResult::ResultCode & result) Line 403	C++
 	EditorLib.dll!AzToolsFramework::Prefab::JsonInstanceSerializer::Load(void * outputValue, const AZ::Uuid & outputValueTypeId, const rapidjson_ly::GenericValue<rapidjson_ly::UTF8<char>,rapidjson_ly::MemoryPoolAllocator<rapidjson_ly::CrtAllocator>> & inputValue, AZ::JsonDeserializerContext & context) Line 205	C++
 	EditorLib.dll!AZ::JsonDeserializer::DeserializerDefaultCheck(AZ::BaseJsonSerializer * serializer, void * object, const AZ::Uuid & typeId, const rapidjson_ly::GenericValue<rapidjson_ly::UTF8<char>,rapidjson_ly::MemoryPoolAllocator<rapidjson_ly::CrtAllocator>> & value, bool isNewInstance, AZ::JsonDeserializerContext & context) Line 35	C++
 	EditorLib.dll!AZ::JsonDeserializer::Load(void * object, const AZ::Uuid & typeId, const rapidjson_ly::GenericValue<rapidjson_ly::UTF8<char>,rapidjson_ly::MemoryPoolAllocator<rapidjson_ly::CrtAllocator>> & value, bool isNewInstance, AZ::JsonDeserializer::UseTypeDeserializer custom, AZ::JsonDeserializerContext & context) Line 56	C++
 	EditorLib.dll!AZ::JsonSerialization::Load(void * object, const AZ::Uuid & objectType, const rapidjson_ly::GenericValue<rapidjson_ly::UTF8<char>,rapidjson_ly::MemoryPoolAllocator<rapidjson_ly::CrtAllocator>> & root, AZ::JsonDeserializerSettings & settings) Line 244	C++
 	[Inline Frame] EditorLib.dll!AZ::JsonSerialization::Load(AzToolsFramework::Prefab::Instance &) Line 331	C++
 	EditorLib.dll!AzToolsFramework::Prefab::PrefabDomUtils::Internal::LoadInstanceHelper(AzToolsFramework::Prefab::Instance & instance, const rapidjson_ly::GenericDocument<rapidjson_ly::UTF8<char>,rapidjson_ly::MemoryPoolAllocator<rapidjson_ly::CrtAllocator>,rapidjson_ly::CrtAllocator> & prefabDom, AzToolsFramework::Prefab::PrefabDomUtils::LoadFlags flags, AZ::JsonDeserializerSettings & settings) Line 166	C++
 	EditorLib.dll!AzToolsFramework::Prefab::PrefabDomUtils::LoadInstanceFromPrefabDom(AzToolsFramework::Prefab::Instance & instance, AZStd::vector<AZ::Entity *,AZStd::allocator> & newlyAddedEntities, const rapidjson_ly::GenericDocument<rapidjson_ly::UTF8<char>,rapidjson_ly::MemoryPoolAllocator<rapidjson_ly::CrtAllocator>,rapidjson_ly::CrtAllocator> & prefabDom, AzToolsFramework::Prefab::PrefabDomUtils::LoadFlags flags) Line 313	C++
 	EditorLib.dll!AzToolsFramework::Prefab::InstanceUpdateExecutor::UpdateTemplateInstancesInQueue() Line 205	C++
`

**Comment by galibzon** [comment]
Added @monroegm to follow up with his team.

**Comment by galibzon** [comment]
@srikappa-amzn , FYI: @gadams3 had issues with `PyIdleWaitFrames` in the past when enabling python scriptability for the Material Editor. See the difference:

**Comment by srikappa-amzn** [comment]
Looking at the error log, I can think of 2 ways to fix the crash:

1. Implement smart reloading logic such that entities are not destroyed and recreated when a new entity is added --- this needs to be done by folks knowledgeable about prefabs
2. Even if an entity is being destroyed and recreated, why is that causing a crash in AZ::Render::EditorMaterialSystemComponentNotifications ? And why does that only happen when there are many entities? This indicates that there is a bug lurking somewhere there ---- this needs to be fixed by folks knowledgeable about Atom code.

**Comment by galibzon** [comment]
@srikappa-amzn , per @gadams3 findings, the fact that it crashes on AZ::Render::EditorMaterialSystemComponentNotifications, is more of a coincidence that it is the last component that is being added to the entity. Same crash should occur if there are other components as the last component.

**Comment by srikappa-amzn** [comment]
@gadams3 , entities getting destroyed and recreated during prefab propagation is very common. So I'm not sure how this can be happening on any 'last added' component on the entities. Maybe I'm missing some connecting pieces here. 

@galibzon , will be happy to get into a call to narrow the root cause down further.

**Comment by forkercat** [comment]
Was able to reproduce it by running the following command in console.

```shell
pyRunFile ..\..\\Levels\goal_9ks_1kd\Scripts\populate_static_objects.py -m 100 -o 0 -c 100
```

Sometimes it crashes sometimes it does work fine. But it usually crashes around 30-60/100.

- I tried disabling the material component and it won't crash even running `-m 100 -c 100` for multiple times.
- I also tried putting collider comp. right after material comp. (then material won't be the last). It still crashes.
